manager/snap.go

```diff
@@ -79,6 +87,126 @@ func (snap *snap) GetProxySettings() (proxy.Settings, error) {
+// possible and an appropriate error will be returned. For such use-cases, the
+// store proxy must be configured via a call to ConfigureStoreProxy instead.
+//
+// If snap also needs to use HTTP/HTTPS proxies to talk to the outside world,
+// these need to be configured separately before invoking this method via a
+// call to SetProxy.
+func (snap *Snap) ConfigureStoreProxyFromURL(proxyURL string) error {
+	u, err := url.Parse(proxyURL)
+	if err != nil {
+		return errors.Annotate(err, "proxy URL not valid")
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return errors.NotValidf("proxy URL scheme %q", u.Scheme)
+	}
+
+	res, err := http.Get(proxyURL + "/v2/auth/store/assertions")
+	if err != nil {
+		return errors.Annotatef(err, "could not contact snap store proxy at %q", proxyURL)
+	}
+	defer func() { _ = res.Body.Close() }()
+	if res.StatusCode != http.StatusOK {
+		return errors.Annotatef(err, "could not retrieve assertions from proxy at %q. If using an air-gapped proxy you must manually provide the assertions file and store ID", proxyURL)
+	}
+
+	data, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return errors.Annotatef(err, "could not read assertions response from proxy at %q", proxyURL)
+	}
+	assertions := string(data)
+	storeID, err := findStoreID(assertions, proxyURL)
+	if err != nil {
+		return err
+	}
+
+	return snap.ConfigureStoreProxy(assertions, storeID)
+}
+
+func findStoreID(assertions string, proxyURL string) (string, error) {
+	var storeID string
+	for _, match := range storeInAssertionRE.FindAllStringSubmatch(assertions, -1) {
+		if len(match) != 3 {
+			continue
+		}
+
+		// Found store assertion but not for the URL provided
+		if match[2] != proxyURL {
+			continue
+		}
+
+		// Found same URL but different store ID
+		if storeID != "" && match[1] != storeID {
+			return "", errors.Errorf("assertions response from proxy at %q is ambiguous as it contains multiple entries with the same proxy URL but different store ID", proxyURL)
+		}
+
+		storeID = match[1]
+	}
+
+	if storeID == "" {
+		return "", errors.NotFoundf("store ID in assertions response from proxy at %q", proxyURL)
+	}
+
+	return storeID, nil
+}
+
+// ConfigureStoreProxy sets up snapd to connect to the snap store proxy
+// instance defined in the provided assertions using the provided store ID.
+//
+// If snap also needs to use HTTP/HTTPS proxies to talk to the outside world,
+// these need to be configured separately before invoking this method via a
+// call to SetProxy.
+func (snap *Snap) ConfigureStoreProxy(assertions, storeID string) error {
+	// Setup proxy based on the instructions from:
+	// https://docs.ubuntu.com/snap-store-proxy/en/devices
+	//
+	// Note that while the above instructions run "snap ack /dev/stdin" the
+	// code below will instead write the assertions to a temp file and pass
+	// that to snap ack. This is purely done to make testing easier.
+	assertFile, err := ioutil.TempFile("", "assertions")
```

Comment:
There doesn't seem to be anything that deletes this tempfile once it has been used.

I'd rather see us use stdin rather than a tempfile, but I can see that RunCommandWithRetry probably doesn't support that.

Reply:
There is a defer below although I accidentally placed it at L181 instead of L177. 

My initial approach created an exec.Cmd and piped the asserts via the stdin reader but it's really hard to test without having to jump through hoops.